### PR TITLE
feat(symbolicator): Ignore all built-in sources based on config option 

### PR DESCRIPTION
**Pull request by relaxolotl**
This allows sentry to omit all built-in sources during symbolication, which should open up an option to mitigate the impact of a cache refresh.

Option registration will be done in a separate PR (#27265), assuming 

> The part where we register the option needs to be a separate PR as a manual step needs to be performed between registering and reading the option.

from https://github.com/getsentry/sentry/pull/25807#issuecomment-831297640 still applies to new options.

### CODE REVIEW

**Comment by flub** [comment]
> Option registration will be done in a separate PR, assuming
> 
> > The part where we register the option needs to be a separate PR as a manual step needs to be performed between registering and reading the option.
> 
> from [#25807 (comment)](https://github.com/getsentry/sentry/pull/25807#issuecomment-831297640) still applies to new options.

Yes, that PR needs to land first though IIRC

File: src/sentry/options/defaults.py [hunk 1]
```diff
@@ -113,6 +113,9 @@
 # The ratio of requests for which the new stackwalking method should be compared against the old one
 register("symbolicator.compare_stackwalking_methods_rate", default=0.0)
 
+# Ignore built-in sources during symbolication
+register("symbolicator.ignore_builtin_sources", default=False)
```
Comment (flub) [inline src/sentry/options/defaults.py:117]: I think this needs to be in a separate PR that we need to merge first.  Once merged we need to ask the right people to do the right thing (probably should use an ops request for this rather than pester the people who can do this directly).  Only then can we submit the rest.

Comment (flub) [reply src/sentry/options/defaults.py:117]: I might be mistaken (and can't find the jira task somehow?  still learning how that works) but I think the requirement is to specify a list of IDs which should be excluded.  So that ops can set the option to e.g. `sentry:electron` and if a project has both `sentry:electron` and `sentry:microsoft` enabled they'll still symbolicate with `sentry:microsoft` enabled.

There already are some options here registered as a "Sequence" so i guess we should be able to copy the syntax from those.

**Comment by untitaker** [comment]
Hey, not sure if you still want my review, but while this surely takes load off of the system, isn't this directly impacting customer's events and stacktrace quality?

**Comment by flub** [comment]
> Hey, not sure if you still want my review, but while this surely takes load off of the system, isn't this directly impacting customer's events and stacktrace quality?

yes it is.  this is an emergency everything-is-on-fire handle, only to be used if it is better to partially break a few customers instead of entirely break everyone.

**Comment by untitaker** [comment]
Ok, right now this is a binary option, I wonder if we can sample or use sentry.killswitches to restrict impact

**Comment by relaxolotl** [comment]
@untitaker sorry about that, I'd initially requested your review because I got the steps backwards (thought it was read first, then register). The PR to register this new option as you've already noticed was what I wanted to get your feedback on, in case the process had changed since May. Your feedback's still welcome in this PR if you notice any impact that I might've missed.

It looks like I'd forgotten to actually link the register PR, so I've updated the description accordingly. It can be found here: #27265. This is rebased off of #27265's changes, so the `register()` bit should land in `master` once/if that merges.

File: src/sentry/options/defaults.py (hunk 1 above)
Comment (relaxolotl) [reply src/sentry/options/defaults.py:117]: Oh, you're right! I missed the part about filtering them out based on their ID, so I'll need to update this to reflect that, thank you. The matching logic required to make this work would mean it'd be a little difficult to use `sentry.killswitches` as-is, though. I'll leave a longer comment that goes over that.

Comment (relaxolotl) [reply src/sentry/options/defaults.py:117]: Actually, what's a little ambiguous to me is whether ops wants to kill _all_ built-in sources with one switch, or if they want to be able to globally kill specific built-in sources by providing a hit list. I interpreted it being the former, but perhaps they're interested in the latter functionality.

**Comment by relaxolotl** [comment]
Regarding using killswitches: It looks like I could use these, although I'm left with a few questions about them. For context, this particular kill switch needs to be applied globally and should be toggleable without requiring a deployment.

Killswitches are all registered as options of `Any` type and have a default of `[]`. `sentry.killswitches` itself, I see that killswitches usually come with a map of conditions, i.e. a `project_id` to apply the switch to, together with some other condition, like a platform or an `event_id`. If I understand things correctly, the switch conditions are loaded from the config yaml so using these won't break the no-deploy requirement. Is that about right?

As mentioned in a comment above, I need to clarify whether ops wants to be able to kill specific built-in sources or if they're interested in blocking all built-in sources. I'll explore what needs to be done in both strategies, though.

### If we want this switch to target only specific built-in sources
This stops being a binary option, and we'll need to use the contents of the option to filter out built-in sources. Probably shouldn't use `sentry.killswitch` here.

### If we want this switch to turn off all built-in sources
I'd be adding in something like this:
```python
"symbolicator.ignore-all-builtin-sources": KillswitchInfo(
      description="Drop all built-in sources used in symbolication",
      fields={
          "is_on": "Whether the kill switch is on or not",
      },
),
```
This feels a little overcomplicated for what the switch is trying to accomplish, but I could be missing some guarantees that `sentry.killswitches` offers in exchange for this complexity. Other kill switches check multiple conditions; This only has one.

In contrast, I see that we do have a few kill switches that don't make use of `sentry.killswitches`: `store.reprocessing-force-disable`, `transaction-events.force-disable-internal-project`, etc. Several of these are simple True/false conditions similar to the one I'd like to add in this PR, which leads me to think that maybe `sentry.killswitches` might not be the correct solution (or those are old kill switches that need to be migrated over to use that).

Thoughts, @untitaker?

**Comment by flub** [comment]
> As mentioned in a comment above, I need to clarify whether ops wants to be able to kill specific built-in sources or if they're interested in blocking all built-in sources. I'll explore what needs to be done in both strategies, though.

Pretty sure it's the latter, from the story at https://getsentry.atlassian.net/browse/NATIVE-55:
> On the sentry side, we want to add an internal use only (operations) configuration option **to filter out** any of the builtin symbol sources (MS, Electron, etc) **by their ID**.

(emphasis mine)

**Comment by relaxolotl** [comment]
Updated this so it filters sources out based on what's set in the config option, so this should be good for a re-review. I've opted not to roll this into `sentry.killswitch` since this isn't a simple binary check.

File: src/sentry/lang/native/symbolicator.py [hunk 2]
```diff
@@ -352,6 +352,12 @@ def get_sources_for_project(project):
 
     sources = []
 
+    def remove_ignored_sources(sources):
```
Comment (flub) [inline src/sentry/lang/native/symbolicator.py:355]: This is all nitpicking, so do with it as you like.
* I think naming-wise this is what is usually known as a filter, so I'd go with `filter_ignored_sources()`.
* This doesn't use any free variables so don't need to be a closure.  I'd make it a top-level function instead of nesting it.  This means you can unittest it easily etc.

File: tests/sentry/lang/native/test_symbolicator.py [hunk 3]
```diff
@@ -105,6 +106,73 @@ def test_sources_custom_disabled(default_project):
...
+    default_project.update_option("sentry:symbol_sources", CUSTOM_SOURCE_CONFIG)
+    options.set("symbolicator.ignored_sources", ["honk"])
+
+    with Feature(features):
+        sources = get_sources_for_project(default_project)
+
+    source_ids = map(lambda s: s["id"], sources)
+    assert source_ids == ["sentry:project", "custom", "sentry:microsoft", "sentry:electron"]
+
+
```
Comment (flub) [inline tests/sentry/lang/native/test_symbolicator.py:175]: yay for tests! ❤️

**Review by flub — APPROVED** [review]

File: src/sentry/lang/native/symbolicator.py (hunk 2 above)
Comment (relaxolotl) [reply src/sentry/lang/native/symbolicator.py]: Sorry for dragging this out; While pulling `remove_ignored_sources` out I realized I'd completely missed something in my implementation: It doesn't acknowledge aliases at all. I figure ops would probably (hopefully) be more likely to use aliases with this kill switch, so I changed this implementation a little.

The new logic recognizes and filters for both aliases and de-aliased internal source IDs and pulls all of the filtering logic out into its own top-level function, making it pretty unit testable. This still gets covered in the integration tests seeing as I'd accidentally broken them in some of my commits, so I think we're good there.

Seeing as I'd reworked the logic quite a bit, could I bother you to take _yet another look_ at this? I'm mostly hoping that you could check to make sure I haven't made any errors handling aliases, and if there's something else I might've missed with source IDs like aliases.

**Comment by relaxolotl** [comment]
Keeping this as draft as this can't be merged until a manual action is completed on prod. This should be ready for review, though.

**Review by Swatinem — APPROVED** [review]

**Comment by untitaker** [comment]
I'm a bit late to this, betty I believe you're right with everything you say, but I think there's a misunderstanding on how killswitches work regardless:

```
"symbolicator.ignore-all-builtin-sources": KillswitchInfo(
      description="Drop all built-in sources used in symbolication",
      fields={
          "is_on": "Whether the kill switch is on or not",
      },
),
```

this doesn't make sense to me, I'd have `source_id` and possibly `project_id` as `fields` instead, i.e. criteria *by which* you want to disable/enable the killswitch by.

Then you can enable/disable the killswitch like so from the CLI:

```
sentry killswitches edit symbolicator.ignore-builtin-sources

<vim opens>

- source_id: 123  # ignore source 123 specifically

# - source_id: ~  # OR ignore all builtin sources

<close vim>

<preview of what you configured + prompt to confirm>
```

File: tests/sentry/lang/native/test_symbolicator.py [hunk 4]
```diff
@@ -265,3 +271,135 @@ def test_reverse_aliases(self, builtin_sources):
         reverse_aliases = symbolicator.reverse_aliases_map(builtin_sources)
         expected = {"sentry:ios-source": "sentry:ios", "sentry:tvos-source": "sentry:ios"}
         assert reverse_aliases == expected
+
+
+class TestIgnoredSourcesFiltering:
```
Comment (untitaker) [inline tests/sentry/lang/native/test_symbolicator.py:276]: fwiw we don't have a lot of pytest-style classes around, may be more consistent to start a new file entirely

File: tests/sentry/lang/native/test_symbolicator.py [hunk 5]
```diff
@@ -265,3 +271,135 @@ def test_reverse_aliases(self, builtin_sources):
...
+        ]
+        builtins.extend(json.loads(CUSTOM_SOURCE_CONFIG))
+        return builtins
+
+    @pytest.fixture
+    def reversed_alias_map(self):
+        return {"sentry:ios-source": "sentry:ios", "sentry:tvos-source": "sentry:ios"}
+
+    # Explicitly empty list of sources
+    @pytest.mark.django_db
```
Comment (untitaker) [inline tests/sentry/lang/native/test_symbolicator.py:309]: If you want to make your testcases faster, I'd remove this marker and use `sentry.testutils.helpers.override_options` ctx mgr since that appears to be the only source of DB access

**Comment by untitaker** [comment]
option has been set in prod so from that end it is safe to merge, fwiw

File: tests/sentry/lang/native/test_symbolicator.py (hunk 5 above)
Comment (flub) [reply tests/sentry/lang/native/test_symbolicator.py]: ooh, this is a great hint!
